misc/server/src/main.rs

```diff
@@ -0,0 +1,213 @@
+        swarm.add_external_address(address.clone())
+    }
+    log::info!(
+        "External addresses: {:?}",
+        swarm.external_addresses().collect::<Vec<_>>()
+    );
+
+    let mut metric_registry = Registry::default();
+    let metrics = Metrics::new(&mut metric_registry);
+    let build_info = Info::new(vec![("version".to_string(), env!("CARGO_PKG_VERSION"))]);
+    metric_registry.register(
+        "build",
+        "A metric with a constant '1' value labeled by version",
+        build_info,
+    );
+    thread::spawn(move || block_on(http_service::metrics_server(metric_registry)));
+
+    let mut bootstrap_timer = Delay::new(BOOTSTRAP_INTERVAL);
+
+    loop {
+        if let Poll::Ready(()) = futures::poll!(&mut bootstrap_timer) {
+            bootstrap_timer.reset(BOOTSTRAP_INTERVAL);
+            let _ = swarm
+                .behaviour_mut()
+                .kademlia
+                .as_mut()
+                .map(|k| k.bootstrap());
+        }
+
+        match swarm.next().await.expect("Swarm not to terminate.") {
+            SwarmEvent::Behaviour(behaviour::BehaviourEvent::Identify(e)) => {
+                info!("{:?}", e);
+                metrics.record(&e);
+
+                if let identify::Event::Received {
+                    peer_id,
+                    info:
+                        identify::Info {
+                            listen_addrs,
+                            protocols,
+                            ..
+                        },
+                } = e
+                {
+                    if protocols.iter().any(|p| *p == kad::PROTOCOL_NAME) {
+                        for addr in listen_addrs {
+                            swarm
+                                .behaviour_mut()
+                                .kademlia
+                                .as_mut()
+                                .map(|k| k.add_address(&peer_id, addr));
+                        }
+                    }
+                }
+            }
+            SwarmEvent::Behaviour(behaviour::BehaviourEvent::Ping(e)) => {
+                debug!("{:?}", e);
+                metrics.record(&e);
+            }
+            SwarmEvent::Behaviour(behaviour::BehaviourEvent::Kademlia(e)) => {
+                debug!("{:?}", e);
+                metrics.record(&e);
+            }
+            SwarmEvent::Behaviour(behaviour::BehaviourEvent::Relay(e)) => {
+                info!("{:?}", e);
+                metrics.record(&e)
+            }
+            SwarmEvent::Behaviour(behaviour::BehaviourEvent::Autonat(e)) => {
+                info!("{:?}", e);
+                // TODO: Add metric recording for `NatStatus`.
+                // metrics.record(&e)
+            }
+            e => {
+                if let SwarmEvent::NewListenAddr { address, .. } = &e {
+                    println!("Listening on {address:?}");
+                }
+
+                metrics.record(&e)
+            }
```

Comment:
Nit: This looks a bit odd, why not pattern match directly right away?

Reply:
Odd indeed. Rational here being that `libp2p-metrics::swarm` tracks all but `SwarmEvent::Behaviour`. These are captured in this match arm.

Refactored with cf79259ba.